### ISSUE TITLE
Question about prompt-free pretraining

### ISSUE THREAD

**Issue by arthur0219** [ultralytics/ultralytics#24661]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, thanks for your great work, and I'm curious about some details about prompt-free pretraining:

1. What training data was used for the prompt-free model pretraining? The pretrained checkpoint ships with 4,585 built-in vocabulary items. Are these 4,585 categories fully covered by Objects365, GQA, and Flickr combined? If some vocabulary items are not found in any of these detection/grounding datasets, how were they handled — were they excluded from training, or were additional datasets used to cover the full vocabulary?

2. During prompt-free model training, are the vocab layer (Linear, 4,585 output channels) and the pf layer (objectness head, Conv2d(c3, 1)) trained simultaneously? Or are they trained in separate stages — e.g., vocab first via linear probing (1 epoch as described in the paper), then pf independently? What is the exact training recipe: which components are trainable, which are frozen, and for how many epochs?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @arthur0219, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your question routed quickly, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
For your prompt-free pretraining question, it would be especially helpful if you can share any specific model, checkpoint, config, paper section, or code reference you are basing the questions on so the team can give the most accurate clarification 🔍
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by arthur0219** [comment]
By the way, this question is for YOLOE-26 specifically

**Comment by raimbekovm** [comment]
@ShuaiLYU

**Comment by ShuaiLYU** [comment]
@arthur0219  , Thx for you question!


- What training data was used for the prompt-free model pretraining? 

YOLOE26 uses the same training data as YOLOE (Objects365, GQA, and Flickr).But we employ a pretrained YOLOE teacher, prompted with the built-in vocabulary of 4585 classes, to refine the train-
ing set. Predicted boxes that are absent from the original annotations, have IoU < 0.5 with any existing ground-truth
box, and exceed a confidence threshold of 0.5 are appended as pseudo-labels for a second round of training. 


- The pretrained checkpoint ships with 4,585 built-in vocabulary items. Are these 4,585 categories fully covered by Objects365, GQA, and Flickr combined? 

Yep, the 4,585 built-in vocabulary items are fully covered after the pseudo-labeling process. 

- If some vocabulary items are not found in any of these detection/grounding datasets, how were they handled — were they excluded from training, or were additional datasets used to cover the full vocabulary?

Theoretically, vocabulary items with similar semantics to those seen during training should be well predicted. You can do testing to see how well the model performs on unseen categories.

- During prompt-free model training, are the vocab layer (Linear, 4,585 output channels) and the pf layer (objectness head, Conv2d(c3, 1)) trained simultaneously? - Or are they trained in separate stages — e.g., vocab first via linear probing (1 epoch as described in the paper), then pf independently? What is the exact training recipe: which components are trainable, which are frozen, and for how many epochs?

The vocab layer and the pf layer are not trained simultaneously.  YOLOE is trained in three stages: a text-prompt stage (TP) , a visual-prompt stage (VP) and a prompt-free stage (PF). For the TP stage, 15-30 epochs for different sizes. Then vocab can be extracted with it. For the PF stage, pretrained weights from the TP stage are loaded, and "object" is taken as text-prompt as a starting point for the PF single-class training, with 10 epochs.

**Comment by arthur0219** [comment]
@ShuaiLYU   Thanks for your reply, and I want to check 2 details with you, please

1. Regarding "the 4,585 built-in vocabulary items are fully covered after the pseudo-labeling process" — does this mean that after the text-prompt (TP) stage is complete, the vocab layer requires no dedicated training? In other words, the 4,585 classes are simply reparameterized by fusing their text embeddings (CLIP + RepRTA → TPE) into the classification convolution, directly producing the Linear(c3, 4585) vocab layer — with no further gradient-based training on that layer?

2. For the prompt-free (PF) single-class training stage: is it effectively a linear probing setup? Specifically, is everything frozen except the pf head (Conv2d(c3, 1)), which is trained using the single text prompt "object"? In this case, does "object" go through the standard CLIP → (RepRTA, if present) → fuse pathway to initialize pf, and then only pf's weights are updated during the 10-epoch PF training?

**Comment by ShuaiLYU** [comment]
-  yep , you understanding is right!    @arthur0219 

-  We follow the YOLO paper: only the cv3[-1] layer is trained, and all other layers are frozen. This ensures that all text-prompt related layers are not updated, which is determined by the separate training of TP and PF. If we can jointly train different modes in the future, we may unfreeze all layers for training.

**Comment by arthur0219** [comment]
Thanks for your reply

**Comment by glenn-jocher** [comment]
You're welcome — glad that clarified it, and if you have any follow-up YOLOE-26 questions, please open a new [Discussion](https://github.com/orgs/ultralytics/discussions) so the Ultralytics team can see it.
